File: controller/routes/chats.py
"""Chat session endpoints."""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, Request

from ..store import ChatStore

router = APIRouter(tags=["Chats"])


def get_chat_store() -> ChatStore:
    """Get chat store instance."""
    from ..app import get_chat_store as _get_chat_store
    return _get_chat_store()


@router.get("/chats")
async def list_chat_sessions(chat_store: ChatStore = Depends(get_chat_store)):
    """List all chat sessions."""
    return chat_store.list_sessions()


@router.get("/chats/{session_id}")
async def get_chat_session(session_id: str, chat_store: ChatStore = Depends(get_chat_store)):
    """Get a chat session with its messages."""
    session = chat_store.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return {"session": session}


@router.post("/chats")
async def create_chat_session(request: Request, chat_store: ChatStore = Depends(get_chat_store)):
    """Create a new chat session."""
    body = await request.json()
    session_id = str(uuid.uuid4())
    title = body.get("title", "New Chat")
    model = body.get("model")
    session = chat_store.create_session(session_id, title, model)
    return {"session": session}


@router.put("/chats/{session_id}")
async def update_chat_session(session_id: str, request: Request, chat_store: ChatStore = Depends(get_chat_store)):
    """Update a chat session (title, model)."""
    body = await request.json()
    if not chat_store.update_session(session_id, body.get("title"), body.get("model")):
        raise HTTPException(status_code=404, detail="Session not found")
    return {"success": True}


@router.delete("/chats/{session_id}")
async def delete_chat_session(session_id: str, chat_store: ChatStore = Depends(get_chat_store)):
    """Delete a chat session."""
    if not chat_store.delete_session(session_id):
        raise HTTPException(status_code=404, detail="Session not found")
    return {"success": True}


@router.post("/chats/{session_id}/messages")
async def add_chat_message(session_id: str, request: Request, chat_store: ChatStore = Depends(get_chat_store)):
    """Add a message to a chat session."""
    body = await request.json()
    message_id = body.get("id", str(uuid.uuid4()))
    role = body.get("role", "user")
    content = body.get("content")
    model = body.get("model")
    tool_calls = body.get("tool_calls")
    request_prompt_tokens = body.get("request_prompt_tokens")
    request_tools_tokens = body.get("request_tools_tokens")
    request_total_input_tokens = body.get("request_total_input_tokens")
    request_completion_tokens = body.get("request_completion_tokens")
    message = chat_store.add_message(
        session_id, message_id, role, content, model, tool_calls,
        request_prompt_tokens, request_tools_tokens, request_total_input_tokens, request_completion_tokens
    )
    return message


@router.get("/chats/{session_id}/usage")
async def get_chat_usage(session_id: str, chat_store: ChatStore = Depends(get_chat_store)):
    """Get token usage for a chat session."""
    return chat_store.get_usage(session_id)


@router.post("/chats/{session_id}/fork")
async def fork_chat_session(session_id: str, request: Request, chat_store: ChatStore = Depends(get_chat_store)):
    """Fork a chat session, optionally from a specific message."""
    body = await request.json()
    new_id = str(uuid.uuid4())
    message_id = body.get("message_id")
    model = body.get("model")
    title = body.get("title")
    session = chat_store.fork_session(session_id, new_id, message_id, model, title)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return {"session": session}

File: controller/src/routes/chats.ts
// CRITICAL
import type { Hono } from "hono";
import { randomUUID } from "node:crypto";
import type { AppContext } from "../types/context";
import { notFound } from "../core/errors";
import { compactChatSession } from "../services/chat-compaction";

/**
 * Generate a simple title from the first few words of a message
 */
function generateTitleFromMessage(content: string): string {
  if (!content || !content.trim()) {
    return "New Chat";
  }

  const cleaned = content
    .replace(/\n/g, " ")
    .replace(/[^\w\s'-]/g, " ")
    .replace(/\s+/g, " ")
    .trim();

  const words = cleaned.split(" ").filter((w) => w.length > 0);
  const titleWords = words.slice(0, 5);

  if (titleWords.length === 0) {
    return "New Chat";
  }

  const title = titleWords
    .map((w, i) => (i === 0 ? w.charAt(0).toUpperCase() + w.slice(1) : w))
    .join(" ");

  return title.length > 50 ? title.slice(0, 47) + "..." : title;
}

/**
 * Register chat session routes.
 * @param app - Hono app.
 * @param context - App context.
 */
export const registerChatsRoutes = (app: Hono, context: AppContext): void => {
  app.get("/chats", async (ctx) => {
    return ctx.json(context.stores.chatStore.listSessions());
  });

  app.get("/chats/:sessionId", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const session = context.stores.chatStore.getSession(sessionId);
    if (!session) {
      throw notFound("Session not found");
    }
    return ctx.json({ session });
  });

  app.post("/chats/:sessionId/compact", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    let body: Record<string, unknown> = {};
    try {
      body = (await ctx.req.json()) as Record<string, unknown>;
    } catch {
      body = {};
    }

    const model = typeof body["model"] === "string" ? body["model"] : undefined;
    const systemPrompt = typeof body["system"] === "string" ? body["system"] : undefined;
    const title = typeof body["title"] === "string" ? body["title"] : undefined;
    const preserveFirst = body["preserve_first"] !== false;
    const preserveLast = body["preserve_last"] !== false;

    const result = await compactChatSession(context, sessionId, {
      ...(model ? { model } : {}),
      ...(systemPrompt ? { systemPrompt } : {}),
      ...(title ? { title } : {}),
      preserveFirst,
      preserveLast,
    });
    return ctx.json(result);
  });

  app.post("/chats", async (ctx) => {
    const body = (await ctx.req.json()) as Record<string, unknown>;
    const sessionId = randomUUID();
    const title = typeof body["title"] === "string" ? body["title"] : "New Chat";
    const model = typeof body["model"] === "string" ? body["model"] : undefined;
    const agentState = body["agent_state"];
    const session = context.stores.chatStore.createSession(sessionId, title, model, undefined, agentState);
    return ctx.json({ session });
  });

  app.put("/chats/:sessionId", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const body = (await ctx.req.json()) as Record<string, unknown>;
    const title = typeof body["title"] === "string" ? body["title"] : undefined;
    const model = typeof body["model"] === "string" ? body["model"] : undefined;
    const hasAgentState = Object.prototype.hasOwnProperty.call(body, "agent_state");
    const agentState = hasAgentState ? body["agent_state"] : undefined;
    const updated = context.stores.chatStore.updateSession(sessionId, title, model, agentState);
    if (!updated) {
      throw notFound("Session not found");
    }
    return ctx.json({ success: true });
  });

  app.delete("/chats/:sessionId", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const deleted = context.stores.chatStore.deleteSession(sessionId);
    if (!deleted) {
      throw notFound("Session not found");
    }
    return ctx.json({ success: true });
  });

  app.post("/chats/:sessionId/messages", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const body = (await ctx.req.json()) as Record<string, unknown>;
    const messageId = typeof body["id"] === "string" ? body["id"] : randomUUID();
    const role = typeof body["role"] === "string" ? body["role"] : "user";
    const content = typeof body["content"] === "string" ? body["content"] : undefined;
    const model = typeof body["model"] === "string" ? body["model"] : undefined;
    const toolCalls = Array.isArray(body["tool_calls"]) ? body["tool_calls"] : undefined;
    const parts = Array.isArray(body["parts"]) ? body["parts"] : undefined;
    const metadata = Object.prototype.hasOwnProperty.call(body, "metadata") ? body["metadata"] : undefined;
    const promptTokens = typeof body["request_prompt_tokens"] === "number" ? body["request_prompt_tokens"] : undefined;
    const toolsTokens = typeof body["request_tools_tokens"] === "number" ? body["request_tools_tokens"] : undefined;
    const totalInputTokens = typeof body["request_total_input_tokens"] === "number" ? body["request_total_input_tokens"] : undefined;
    const completionTokens = typeof body["request_completion_tokens"] === "number" ? body["request_completion_tokens"] : undefined;

    const message = context.stores.chatStore.addMessage(
      sessionId,
      messageId,
      role,
      content,
      model,
      toolCalls,
      promptTokens,
      toolsTokens,
      totalInputTokens,
      completionTokens,
      parts,
      metadata,
    );
    return ctx.json(message);
  });

  app.get("/chats/:sessionId/usage", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    return ctx.json(context.stores.chatStore.getUsage(sessionId));
  });

  app.post("/chats/retitle-all", async (ctx) => {
    const sessions = context.stores.chatStore.listSessions();
    let updated = 0;
    let skipped = 0;

    for (const session of sessions) {
      const sessionId = String(session["id"]);
      const fullSession = context.stores.chatStore.getSession(sessionId);
      if (!fullSession) {
        skipped++;
        continue;
      }

      const messages = (fullSession["messages"] ?? []) as Array<Record<string, unknown>>;
      const firstUserMessage = messages.find((m) => m["role"] === "user");

      if (!firstUserMessage || !firstUserMessage["content"]) {
        skipped++;
        continue;
      }

      const newTitle = generateTitleFromMessage(String(firstUserMessage["content"]));
      context.stores.chatStore.updateSession(sessionId, newTitle);
      updated++;
    }

    return ctx.json({ updated, skipped, total: sessions.length });
  });

  app.post("/chats/:sessionId/fork", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const body = (await ctx.req.json()) as Record<string, unknown>;
    const newId = randomUUID();
    const messageId = typeof body["message_id"] === "string" ? body["message_id"] : undefined;
    const model = typeof body["model"] === "string" ? body["model"] : undefined;
    const title = typeof body["title"] === "string" ? body["title"] : undefined;
    const session = context.stores.chatStore.forkSession(sessionId, newId, messageId, model, title);
    if (!session) {
      throw notFound("Session not found");
    }
    return ctx.json({ session });
  });
};
